internal/database/types_hcpcluster.go

```diff
@@ -0,0 +1,91 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package database
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/Azure/ARO-HCP/internal/api"
+)
+
+type CustomerDesiredHCPClusterState struct {
+	// HCPOpenShiftCluster contains the desired state from a customer.  It is filtered to only those fields that customers
+	// are able to set.
+	// We will whitelist specific fields as we go through to prevent conflicts.
+	HCPOpenShiftCluster api.HCPOpenShiftCluster `json:"hcpOpenShiftCluster"`
+}
+
+type ServiceProviderHCPClusterState struct {
+	// HCPOpenShiftCluster contains the service provider owned state.  It is filtered to only those fields that the service provider owns.
+	// We will whitelist specific fields as we go through to prevent conflicts.
+	HCPOpenShiftCluster api.HCPOpenShiftCluster `json:"hcpOpenShiftCluster"`
+}
+
+var FilterHCPClusterState ResourceDocumentStateFilter = newJSONRoundTripFilterer(
+	func() any { return &CustomerDesiredHCPClusterState{} },
+	func() any { return &ServiceProviderHCPClusterState{} },
+)
+
+type jsonRoundTripFilterer struct {
+	newCustomerDesiredStateFn func() any
+	newServiceProviderStateFn func() any
+}
+
+func newJSONRoundTripFilterer(newCustomerDesiredStateFn, newServiceProviderStateFn func() any) ResourceDocumentStateFilter {
+	return jsonRoundTripFilterer{
+		newCustomerDesiredStateFn: newCustomerDesiredStateFn,
+		newServiceProviderStateFn: newServiceProviderStateFn,
+	}
+}
+
+func (r jsonRoundTripFilterer) RemoveUnknownFields(toMutate *ResourceDocument) error {
+	filteredCustomerDesiredState, err := superExpensiveButSimpleRoundFilterForUnknownFields(toMutate.CustomerDesiredState, r.newCustomerDesiredStateFn())
+	if err != nil {
+		return err
+	}
+	toMutate.CustomerDesiredState = filteredCustomerDesiredState
+
+	filteredServiceProviderState, err := superExpensiveButSimpleRoundFilterForUnknownFields(toMutate.ServiceProviderState, r.newCustomerDesiredStateFn())
+	if err != nil {
+		return err
+	}
+	toMutate.ServiceProviderState = filteredServiceProviderState
+
+	return nil
+}
+
+func superExpensiveButSimpleRoundFilterForUnknownFields(startingMap map[string]any, filterObj any) (map[string]any, error) {
+	if len(startingMap) == 0 {
+		return startingMap, nil
+	}
+	currBytes, err := json.Marshal(startingMap)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal original: %w", err)
+	}
+	if err := json.Unmarshal(currBytes, filterObj); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal into filterObj: %w", err)
+	}
+	filteredBytes, err := json.Marshal(filterObj)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal filterObj: %w", err)
+	}
+	filteredMap := map[string]any{}
+	if err := json.Unmarshal(filteredBytes, &filteredMap); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal into filtered map: %w", err)
+	}
+
+	return filteredMap, nil
+}
```

Comment:
I'm still getting my head around this and don't fully know what you have in mind for it, but just a thought:

I recently started using [github.com/evanphx/json-patch](https://pkg.go.dev/github.com/evanphx/json-patch) to properly apply customer PATCH request bodies and I wonder if that could be beneficial here.

In other words, maybe write `filterObj` as a JSON Merge Patch and apply it with [jsonpatch.MergePatch](https://pkg.go.dev/github.com/evanphx/json-patch#MergePatch).  Don't know that it would be any less expensive but maybe could simplify the design?